compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1964,6 +1964,10 @@ impl<'tcx> Rvalue<'tcx> {
             // <https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html>
             Rvalue::Cast(CastKind::PointerExposeAddress, _, _) => false,
 
+            // Keep transmutes around because they might be UB, and we want the
+            // option to catch those at runtime (at least in debug/non-opt).
+            Rvalue::Cast(CastKind::Transmute, _, _) => false,
+
```

Comment:
Why? Removing UB is fine with optimizations. If UB must not be removed, then optimizations should be disabled like miri does.

Reply:
Hmm, looking again, I'm not sure why I thought this was the best option.  I'll test the case using custom MIR instead, and should be fine.